File: Go/MinStack/MinStack.go
package MinStack

import "stack"
import "errors"

type MinStack struct {
	mainStack *stack.Stack
	minStack  *stack.Stack
}

func (st *MinStack) Init() {
	st.mainStack = new(stack.Stack)
	st.minStack = new(stack.Stack)
}

func (st *MinStack) Push(x int) {
	st.mainStack.Push(x)
	if (!st.minStack.IsEmpty() && st.minStack.Peek().(int) >= x) ||
		st.minStack.IsEmpty() {
		st.minStack.Push(x)
	}
}

func (st *MinStack) Pop() (res int, err error) {
	if st.mainStack.IsEmpty() || st.minStack.IsEmpty() {
		return 0, errors.New("Stack is empty")
	}
	res = st.mainStack.Pop().(int)
	if res == st.minStack.Peek().(int) {
		st.minStack.Pop()
	}
	return res, nil
}

func (st *MinStack) Top() (res int, err error) {
	if st.minStack.IsEmpty() {
		return 0, errors.New("Stack is empty")
	}
	return st.mainStack.Peek().(int), nil
}

func (st *MinStack) GetMin() (res int, err error) {
	if st.minStack.IsEmpty() {
		return 0, errors.New("Stack is empty")
	}
	return st.minStack.Peek().(int), nil
}

File: Scala/src/MinStack/MinStack.scala
package MinStack

import scala.collection.mutable.Stack

class MinStack {
  private val mainStack: Stack[Int] = Stack()
  private val minStack: Stack[Int] = Stack()

  def push(x: Int) = {
    mainStack.push(x)
    if ((!minStack.isEmpty && minStack.top >= x) || minStack.isEmpty) {
      minStack.push(x)
    }
  }

  def pop() = {
    assert(!mainStack.isEmpty)
    val peek: Int = mainStack.pop()
    if (!minStack.isEmpty && peek == minStack.top) {
      minStack.pop()
    }
  }

  def top() = {
    assert(!mainStack.isEmpty)
    mainStack.top
  }

  def getMin() = {
    assert(!minStack.isEmpty)
    minStack.top
  }
}
